beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -0,0 +1,1307 @@
+    pub fn head_beacon_block_root(&self) -> Hash256 {
+        self.canonical_head
+            .cached_head_read_lock()
+            .snapshot
+            .beacon_block_root
+    }
+
+    /// Returns the slot of the highest block in the canonical chain.
+    pub fn best_slot(&self) -> Slot {
+        self.canonical_head
+            .cached_head_read_lock()
+            .snapshot
+            .beacon_block
+            .slot()
+    }
+
+    /// Returns a `Arc` of the `BeaconSnapshot` at the head of the canonical chain.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_snapshot(&self) -> Arc<BeaconSnapshot<T::EthSpec>> {
+        self.canonical_head.cached_head_read_lock().snapshot.clone()
+    }
+
+    /// Returns the beacon block at the head of the canonical chain.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_beacon_block(&self) -> Arc<SignedBeaconBlock<T::EthSpec>> {
+        self.head_snapshot().beacon_block.clone()
+    }
+
+    /// Returns a clone of the beacon state at the head of the canonical chain.
+    ///
+    /// Cloning the head state is expensive and should generally be avoided outside of tests.
+    ///
+    /// See `Self::head` for more information.
+    pub fn head_beacon_state_cloned(&self) -> BeaconState<T::EthSpec> {
+        // Don't clone whilst holding the read-lock, take an Arc-clone to reduce lock contention.
+        let snapshot: Arc<_> = self.head_snapshot();
+        snapshot
+            .beacon_state
+            .clone_with(CloneConfig::committee_caches_only())
+    }
+
+    /// Execute the fork choice algorithm and enthrone the result as the canonical head.
+    ///
+    /// This method replaces the old `BeaconChain::fork_choice` method.
+    pub async fn recompute_head_at_current_slot(self: &Arc<Self>) -> Result<(), Error> {
+        let current_slot = self.slot()?;
+        self.recompute_head_at_slot(current_slot).await
+    }
+
+    /// Execute the fork choice algorithm and enthrone the result as the canonical head.
+    ///
+    /// The `current_slot` is specified rather than relying on the wall-clock slot. Using a
+    /// different slot to the wall-clock can be useful for pushing fork choice into the next slot
+    /// *just* before the start of the slot. This ensures that block production can use the correct
+    /// head value without being delayed.
+    pub async fn recompute_head_at_slot(self: &Arc<Self>, current_slot: Slot) -> Result<(), Error> {
+        metrics::inc_counter(&metrics::FORK_CHOICE_REQUESTS);
+        let _timer = metrics::start_timer(&metrics::FORK_CHOICE_TIMES);
+
+        let chain = self.clone();
+        match self
+            .task_executor
+            .spawn_blocking_handle(
+                move || chain.recompute_head_at_slot_internal(current_slot),
+                "recompute_head_internal",
+            )
+            .ok_or(Error::RuntimeShutdown)?
+            .await
+            .map_err(Error::TokioJoin)?
```

Comment:
I knew I'd get u eventually :p

```suggestion
            .spawn_blocking_handle(
                move || chain.recompute_head_at_slot_internal(current_slot),
                "recompute_head_internal",
            )
            .await?
```

Reply:
Ooh nice one! :1st_place_medal: